File: pki.py
import os
import base64
from cryptography.hazmat.primitives import serialization, hashes
from cryptography.hazmat.primitives.asymmetric import ec, rsa, padding
import pickle

class PKI:
    FILE = 'pki_keys.bin'
    CURVE = ec.SECP256R1
    RSA_BITS = 2048

    @classmethod
    def init(cls):
        if os.path.exists(cls.FILE):
            with open(cls.FILE, 'rb') as f:
                keys = pickle.load(f)
            cls.signing_priv = serialization.load_pem_private_key(keys['signing_priv'], password=None)
            cls.signing_pub = serialization.load_pem_public_key(keys['signing_pub'])
            cls.encrypting_priv = serialization.load_pem_private_key(keys['encrypting_priv'], password=None)
            cls.encrypting_pub = serialization.load_pem_public_key(keys['encrypting_pub'])
        else:
            signing_priv = ec.generate_private_key(curve=cls.CURVE)
            signing_pub = signing_priv.public_key()
            encrypting_priv = rsa.generate_private_key(public_exponent=65537, key_size=cls.RSA_BITS)
            encrypting_pub = encrypting_priv.public_key()

            signing_priv_pem = signing_priv.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption()
            )
            signing_pub_pem = signing_pub.public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo
            )
            encrypting_priv_pem = encrypting_priv.private_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PrivateFormat.PKCS8,
                encryption_algorithm=serialization.NoEncryption()
            )
            encrypting_pub_pem = encrypting_pub.public_bytes(
                encoding=serialization.Encoding.PEM,
                format=serialization.PublicFormat.SubjectPublicKeyInfo
            )
            keys = {
                'signing_priv': signing_priv_pem,
                'signing_pub': signing_pub_pem,
                'encrypting_priv': encrypting_priv_pem,
                'encrypting_pub': encrypting_pub_pem
            }
            with open(cls.FILE, 'wb') as f:
                pickle.dump(keys, f)
            cls.signing_priv = serialization.load_pem_private_key(signing_priv_pem, password=None)
            cls.signing_pub = serialization.load_pem_public_key(signing_pub_pem)
            cls.encrypting_priv = serialization.load_pem_private_key(encrypting_priv_pem, password=None)
            cls.encrypting_pub = serialization.load_pem_public_key(encrypting_pub_pem)

    @classmethod
    def sign(cls, message):
        if not hasattr(cls, 'signing_priv'):
            raise ValueError('Keys not initialized')
        signature = cls.signing_priv.sign(
            message.encode(),
            ec.ECDSA(hashes.SHA256())
        )
        return base64.b64encode(signature).decode()

    @classmethod
    def verify(cls, message, signature_b64):
        if not hasattr(cls, 'signing_pub'):
            raise ValueError('Keys not initialized')
        signature = base64.b64decode(signature_b64)
        try:
            cls.signing_pub.verify(
                signature,
                message.encode(),
                ec.ECDSA(hashes.SHA256())
            )
            return True
        except:
            return False

    @classmethod
    def encrypt(cls, message):
        if not hasattr(cls, 'encrypting_pub'):
            raise ValueError('Keys not initialized')
        ciphertext = cls.encrypting_pub.encrypt(
            message.encode(),
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None
            )
        )
        return base64.b64encode(ciphertext).decode()

    @classmethod
    def decrypt(cls, ciphertext_b64):
        if not hasattr(cls, 'encrypting_priv'):
            raise ValueError('Keys not initialized')
        ciphertext = base64.b64decode(ciphertext_b64)
        plaintext = cls.encrypting_priv.decrypt(
            ciphertext,
            padding.OAEP(
                mgf=padding.MGF1(algorithm=hashes.SHA256()),
                algorithm=hashes.SHA256(),
                label=None
            )
        )
        return plaintext.decode()

File: pki.cs
using System;
using System.IO;
using System.Security.Cryptography;
using System.Text;

public static class PKI
{
    private static readonly string FILE = "pki_keys.bin";
    private static ECDsa? signingPriv;
    private static ECDsa? signingPub;
    private static RSA? encryptingPriv;
    private static RSA? encryptingPub;

    public static void Init()
    {
        if (File.Exists(FILE))
        {
            byte[] data = File.ReadAllBytes(FILE);
            using var ms = new MemoryStream(data);
            using var br = new BinaryReader(ms);

            // Load signing private key (PKCS8)
            int len = br.ReadInt32();
            byte[] privEcBytes = br.ReadBytes(len);
            signingPriv = ECDsa.Create();
            signingPriv.ImportPKCS8PrivateKey(privEcBytes, out _);

            // Load signing public key (SubjectPublicKeyInfo)
            len = br.ReadInt32();
            byte[] pubEcBytes = br.ReadBytes(len);
            signingPub = ECDsa.Create();
            signingPub.ImportSubjectPublicKeyInfo(pubEcBytes, out _);

            // Load encrypting private key (PKCS8)
            len = br.ReadInt32();
            byte[] privRsaBytes = br.ReadBytes(len);
            encryptingPriv = RSA.Create();
            encryptingPriv.ImportPKCS8PrivateKey(privRsaBytes, out _);

            // Load encrypting public key (SubjectPublicKeyInfo)
            len = br.ReadInt32();
            byte[] pubRsaBytes = br.ReadBytes(len);
            encryptingPub = RSA.Create();
            encryptingPub.ImportSubjectPublicKeyInfo(pubRsaBytes, out _);
        }
        else
        {
            // Generate signing keypair (ECDSA P-256)
            signingPriv = ECDsa.Create(ECCurve.NamedCurves.nistP256);
            signingPub = ECDsa.Create();
            signingPub.ImportSubjectPublicKeyInfo(signingPriv.ExportSubjectPublicKeyInfo(), out _);

            // Generate encrypting keypair (RSA 2048)
            encryptingPriv = RSA.Create(2048);
            encryptingPub = RSA.Create();
            encryptingPub.ImportSubjectPublicKeyInfo(encryptingPriv.ExportSubjectPublicKeyInfo(), out _);

            // Save to file
            using var fs = new FileStream(FILE, FileMode.Create);
            using var bw = new BinaryWriter(fs);

            // Save signing private
            byte[] signingPrivBytes = signingPriv.ExportPKCS8PrivateKey();
            bw.Write(signingPrivBytes.Length);
            bw.Write(signingPrivBytes);

            // Save signing public
            byte[] signingPubBytes = signingPub.ExportSubjectPublicKeyInfo();
            bw.Write(signingPubBytes.Length);
            bw.Write(signingPubBytes);

            // Save encrypting private
            byte[] encryptingPrivBytes = encryptingPriv.ExportPKCS8PrivateKey();
            bw.Write(encryptingPrivBytes.Length);
            bw.Write(encryptingPrivBytes);

            // Save encrypting public
            byte[] encryptingPubBytes = encryptingPub.ExportSubjectPublicKeyInfo();
            bw.Write(encryptingPubBytes.Length);
            bw.Write(encryptingPubBytes);
        }
    }

    public static string Sign(string message)
    {
        ArgumentNullException.ThrowIfNull(signingPriv);
        byte[] data = Encoding.UTF8.GetBytes(message);
        byte[] signature = signingPriv.SignData(data, HashAlgorithmName.SHA256);
        return Convert.ToBase64String(signature);
    }

    public static bool Verify(string message, string signatureB64)
    {
        ArgumentNullException.ThrowIfNull(signingPub);
        byte[] data = Encoding.UTF8.GetBytes(message);
        byte[] signature = Convert.FromBase64String(signatureB64);
        return signingPub.VerifyData(data, signature, HashAlgorithmName.SHA256);
    }

    public static string Encrypt(string message)
    {
        ArgumentNullException.ThrowIfNull(encryptingPub);
        byte[] data = Encoding.UTF8.GetBytes(message);
        byte[] ciphertext = encryptingPub.Encrypt(data, RSAEncryptionPadding.OaepSHA256);
        return Convert.ToBase64String(ciphertext);
    }

    public static string Decrypt(string ciphertextB64)
    {
        ArgumentNullException.ThrowIfNull(encryptingPriv);
        byte[] ciphertext = Convert.FromBase64String(ciphertextB64);
        byte[] plaintext = encryptingPriv.Decrypt(ciphertext, RSAEncryptionPadding.OaepSHA256);
        return Encoding.UTF8.GetString(plaintext);
    }
}

class Program
{
    static void Main()
    {
        PKI.Init();
        string sig = PKI.Sign("Hello World");
        Console.WriteLine("Signature: " + sig);
        Console.WriteLine("Verify: " + PKI.Verify("Hello World", sig)); // True
        string enc = PKI.Encrypt("Secret message");
        Console.WriteLine("Encrypted: " + enc);
        Console.WriteLine("Decrypted: " + PKI.Decrypt(enc)); // Secret message
    }
}
